server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1589,6 +1509,148 @@ static String resolveExpression(String expression, LongSupplier getTime) {
+            IndexAbstraction indexAbstraction = context.getState().getMetadata().getIndicesLookup().get(expression);
+            if (indexAbstraction == null) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    throw notFoundException(expression);
+                }
+            }
+            // treat aliases as unavailable indices when ignoreAliases is set to true (e.g. delete index and update aliases api)
+            if (indexAbstraction.getType() == Type.ALIAS && context.getOptions().ignoreAliases()) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    throw aliasesNotSupportedException(expression);
+                }
+            }
+            if (indexAbstraction.isDataStreamRelated() && context.includeDataStreams() == false) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    IndexNotFoundException infe = notFoundException(expression);
+                    // Allows callers to handle IndexNotFoundException differently based on whether data streams were excluded.
+                    infe.addMetadata(EXCLUDED_DATA_STREAMS_KEY, "true");
+                    throw infe;
+                }
+            }
+            return true;
+        }
+
+        private static void validateAliasOrIndex(ExpressionIterable.Expression expression) {
+            if (expression.isExclusion()) {
+                return;
+            }
+            if (Strings.isEmpty(expression.toString())) {
+                throw notFoundException(expression.toString());
+            }
+            // Expressions can not start with an underscore. This is reserved for APIs. If the check gets here, the API
+            // does not exist and the path is interpreted as an expression. If the expression begins with an underscore,
+            // throw a specific error that is different from the [[IndexNotFoundException]], which is typically thrown
+            // if the expression can't be found.
+            if (expression.toString().charAt(0) == '_') {
+                throw new InvalidIndexNameException(expression.toString(), "must not start with '_'.");
+            }
+        }
+    }
+
+    public static final class ExpressionIterable implements Iterable<ExpressionIterable.Expression> {
+        private final List<String> expressions;
+        private final int indexOfFirstWildcard;
+
+        public final class Expression {
+            private final int idx;
+
+            Expression(int idx) {
+                assert idx >= 0;
+                assert idx < ExpressionIterable.this.expressions.size();
+                this.idx = idx;
+            }
+
+            @Override
+            public String toString() {
+                if (isExclusion()) {
+                    // drop the leading "-" if exclusion
+                    return get().substring(1);
```

Comment:
I think overall it's easier to automatically drop the leading "-" when working with expressions that contain exclusions. But I admit it's not great API.

Reply:
I would prefer to not rely toString() for any processing concerns here and only use toString() for logging/debugging purposes. maybe a getAsString or get vs. getRaw or the like.

Reply:
Okay, I can do that.

Reply:
While addressing other comments I have pushed https://github.com/elastic/elasticsearch/pull/92518/commits/2f4aa15b05e7dd47f9eabea4c5246aa48322bab5 which also addresses this point.
There is an `Expression#get` method that returns the expression item's string value without the leading "-" in case it is an exclusion.